### ISSUE TITLE
identifying and referencing grids

### ISSUE THREAD

**Issue by d70-t** [ugrid-conventions/ugrid-conventions#59]
First of all, thanks for the effort on standardizing how grids might be encoded.

In most settings I came across, grid definitions are separated from model output data, because otherwise grid definitions would be repeated on and on again. In consequence, given some model output data, a user must find the corresponding grid definition prior to using the data. This often turns out to be a major challenge and usually is not handled in a common way across models. That's what made me think about ugrid...

So, I'm wondering if there are already efforts / conventions in putting identifiers to a grid (I think of UUIDs or content identifiers), which could be used:
* to make a reference from model output data (or even input data) to a grid
* given two datasets, to quickly check if the data is defined on the same grid (e.g. in order to compute differences)
* to uniquely dentify grids independent of storage technologies / datacenters / access protocols

I know that in GRIB2, there's e.g. a UUID of horizontal grid which could be specified in e.g. Grid definition template 3.101 (e.g. on page 55 of [the WMO Manual on Codes, Annex II](https://library.wmo.int/doc_num.php?explnum_id=10722)). Even if I'm not a particular friend of the GRIB format, using a UUID for this purpose seems to be a great choice (at least as opposed to e.g. a local path to a file), because this enables identification independent on a local computing environment and could work even if data and grids are transferred between computing centers.

The closest I've found in CF conventions are [external Variables](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#external-variables) but there, it's explicitly stated that means of identifying external files are not specified.

**Comment by clyne** [comment]
Injecting something like a UUID into the grid specification and data files seems  pretty straightforward, and would certainly make sanity checks between data and grid files possible. It might very well be worth considering this either as an extension to the UGRID specification, or by proposing it for inclusion in the spec. However, maybe I'm missing something but I'm not seeing how this helps to "find the corresponding grid definition prior to using the data". You still need to know where the grid file lives (e.g. on a local  or remote file system somewhere, on AWS cloud storage, etc.). This is a problem that we often face. Perhaps in addition to a UUID having an attribute that provides a path-like string akin to what [fsspec](https://filesystem-spec.readthedocs.io/en/latest/index.html) understands would be helpful? 

cc @ andrewgettelman

**Comment by ChrisBarker-NOAA** [comment]
This is an important question -- but one I think really belongs to CF, rather than UGRID -- I'm pretty sure that there has been discussion of similar issues recently. And in anycase, we'd want to do the same thing as CF if possible.

@clyne: Good point, I think the goals point to two new things:

- to make a reference from model output data (or even input data) to a grid

That would require some url or fspec of some sort.

- given two datasets, to quickly check if the data is defined on the same grid (e.g. in order to compute differences)

There a UUID would be very helpful

Another option would be a hash of the entire grid -- but maybe that would be hard to keep consistent.

- to uniquely identify grids independent of storage technologies / datacenters / access protocols 

there UUID is good too -- and maybe make the hash idea impractical -- e.g. different store techs might Affect the hash.

**Comment by clyne** [comment]
You are right @ChrisBarker-NOAA. This problem isn't unique to unstructured grids. The CF conventions are the right place for this. Anyone want to suggest an extension to the CF, or have any experience doing it?

**Comment by d70-t** [comment]
The reason why I'd think a UUID helps with finding the grid is: grids (and to some extent model output) are being moved around between computers and datacenters. But still, model output should reference the "same" grid (independent of where it's stored at the moment of access). A path to a local filesystem isn't very good for this purpose, because it doesn't exist on another machine. A URL (e.g. http or the like) also doesn't help too much, because some datacenters don't connect their compute nodes to the internet and usually one doesn't want to re-download the grid all the time to check if it's still the same (e.g. we need some form of identifier which tries to encode immutability). If we'd use a UUID (which seems to be more associated with the idea of identifying a thing than a location), we can still talk about the same thing, indepentent of where it is currently located (or where the closest copy is). Although, it still requires trust that no one messes around with the UUIDs.

Of course, this still requires to have a table which maps from UUID to storage location, but this table is easy to generate and we explicitly want to have one table per datacenter (and maybe another one with publicly accessible locations), such that one can always get the "closest copy". One could use that table as a basis e.g. for an `fsspec` filesystem which uses `UUIDs` as pathes, or as a basis for an [intake](https://intake.readthedocs.io/) catalog, which would also allow to have the same grid stored in different formats across datacenters.

Even better (== no trust required) would be a system of hash-based content-identifiers, which would formally verify that grids are indeed unchanged. This could also simplify moving around and caching of (grid-) datasets between datacenters if requested datasets aren't available. zarr-developers/zarr-specs#82 might be relevant here. A possible implementation of content addressable storage below e.g. zarr (or in principle also netCDF) would be based on [IPLD](https://ipld.io/docs/) and [IPFS](https://ipfs.io/), which in turn can already be used [via fsspec](https://github.com/fsspec/ipfsspec/#ipfsspec). But as mentioned, hashes have a couple of things which may make them a bit too inflexible. I hope we'll get there, but there's still some progress to be made.

I've once made an update to CF, but I don't yet have a clear picture of how this one should look like. We maybe wouldn't just want to toss in a bunch of UUIDs, but we'd have to associate them with some context. Things like:
* this dataset can be identified by that UUID
* this dataset refers to another dataset (identified by the following UUID) for use as horizontal grid
* this dataset refers to another dataset (identified by the following UUID) for use as vertical grid
* this dataset refers to another dataset (identified by the following UUID) for use as 3D grid
* this dataset refers to another dataset (identified by the following UUID) for merging variables into the currently opened dataset

I don't yet have a clear picture of how to encode these cases and if that's a complete (or at least sufficiently relevant) set of use cases. Currently those use-casess are in my mind a bit tied to grids, but yes, those could likely be generalized and then they should go into CF.

**Comment by ChrisBarker-NOAA** [comment]
I've just brought this up on the CF project:

https://github.com/cf-convention/cf-conventions/issues/357

Let's see if there is interest there -- as that's where this really belongs.

**Comment by ChrisBarker-NOAA** [comment]
As for your notes:

Yes, I think a UUID or Hash would be important -- but also a url or something would be good.

One question is: 

"how do I find the grid file?"

and another is:

"is the the right grid file?"

As you say, there could be some sort of catalogue that maps UUIDs to actual urls. IN that case, the "where can I find it" url could be a pointer to the catalogue, rather than the file itself.

These urls would indeed be fragile, but I think still useful.

I also really like the Idea of a hash, and a hash could be seen as an alternative to a UUID -- still uniquely identify the. grid, but more robust and verifiable -- if we can come up with a consistent way to hash a grid that may be stored in different file formats, etc.

**Comment by clyne** [comment]
> I've just brought this up on the CF project:
> 
> [cf-convention/cf-conventions#357](https://github.com/cf-convention/cf-conventions/issues/357)
> 
> Let's see if there is interest there -- as that's where this really belongs.

Thanks for bringing this issue up with the CF, @ChrisBarker-NOAA. I think that you have articulated the two separate needs well. CC @brianpm
